File: streamlit.cpp
#define _WIN32_WINNT 0x0A00
#define WINVER 0x0A00

// run g++ blockchain.cpp utilities.cpp streamlit.cpp -o frontend -l SplashKit -l ws2_32 -l sqlite3 -pthread
// Ensure winsock2.h is loaded first (before windows.h)
#include <winsock2.h>
#include <ws2tcpip.h>
#include <windows.h>
#include <sqlite3.h>
#undef ERROR
#undef HELP_KEY

#include "splashkit.h"
#include "httplib.h"
#include "blockchain.h"
#include "utilities.h"
#include "nlohmann/json.hpp"
#include "rsa.hpp"

#include <iostream>
#include <vector>
#include <string>
#include <mutex>
#include <cstdint>
using nlohmann_json = nlohmann::json;
using namespace std;
using namespace rsa;

Blockchain B;
// --- Thread locks for concurrency safety ---
std::mutex blockchain_mutex;
std::mutex user_mutex;

int main()
{
    readChainFromDB(B);

    if (B.chain.empty())
    {
        std::cout << "⚠️ No blockchain data found in DB. Creating genesis block..." << std::endl;
        B.creategenesisBlock();
        saveChainToDB(B);
    }
    httplib::Server svr;

    svr.Post("/create_wallet", [](const httplib::Request &req, httplib::Response &res)
             {
        try {
            auto body = nlohmann_json::parse(req.body);
            string username = body["username"];
 
            Wallet user;
            user.username = username;
            user.privateKey = 0;
            user.publicKey = 0;
            user.modulus = 0;
            generate_keys(user);

            {
                std::lock_guard<std::mutex> lock(user_mutex);
                users.push_back(user);
            }

            nlohmann_json reply = {
                {"publicKey", to_string(user.publicKey)},
                {"privateKey", to_string(user.privateKey)}
            };

            res.set_header("Connection", "keep-alive");
            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_content(reply.dump(), "application/json");
        } catch (const std::exception &e) {
            res.status = 400;
            res.set_content(string("{\"error\":\"") + e.what() + "\"}", "application/json");
        } });

    // --- Mine block endpoint ---
    svr.Post("/mine", [](const httplib::Request &req, httplib::Response &res)
             {
    try {
        auto body = nlohmann_json::parse(req.body);

        uint64_t publicKey = 0;
        try {
            publicKey = body["publicKey"].get<uint64_t>();
        } catch (...) {
            publicKey = std::stoull(body["publicKey"].get<std::string>());
        }

        string prevHash = body["prevHash"];
        string currHash = body["currHash"];

        bool found = false;

        {
            std::lock_guard<std::mutex> lock_user(user_mutex);
            std::lock_guard<std::mutex> lock_block(blockchain_mutex);

            for (auto &u : users) {
                if (u.publicKey == publicKey) {
                    u.mine_block(B, publicKey, prevHash, currHash);
                    found = true;
                    break;
                }
            }
        }

        nlohmann_json reply;
        if (found) {
            reply = {{"status", "mined"}};
            cout << "Mined block added to chain, new length: " << B.chain.size() << endl;

        } else {
            reply = {{"status", "failed"}, {"reason", "User not found"}};
        }

        res.set_header("Connection", "keep-alive");
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_content(reply.dump(), "application/json");
    } 
    catch (const std::exception &e) {
        res.status = 400;
        res.set_content(
            nlohmann_json({{"error", e.what()}}).dump(),
            "application/json"
        );
    } });

    svr.Post("/automine", [](const httplib::Request &req, httplib::Response &res)
             {
        try {
            auto body = nlohmann_json::parse(req.body);
            uint64_t publicKey = body["publicKey"];

            lock_guard<mutex> lock_user(user_mutex);
            lock_guard<mutex> lock_block(blockchain_mutex);

            string prevHash = B.getLatestBlock().previousHash;
            string currHash = B.getLatestBlock().Hash;
            
                
                bool userFound = false;
                for (auto &u : users) {
                    if (u.publicKey == publicKey) {
                     u.mine_block(B, publicKey, prevHash, currHash);
                    userFound = true;
                    break;
                }
            }
        

            if (!userFound) {
                res.status = 404;
                res.set_content("{\"error\":\"User not found\"}", "application/json");
                return;
            }

            nlohmann_json reply = {
                {"status", "auto mined"},
                {"prevHash", prevHash},
                {"currHash", currHash}
            };

            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_content(reply.dump(), "application/json");
        } catch (const std::exception &e) {
            res.status = 400;
            res.set_content(string("{\"error\":\"") + e.what() + "\"}", "application/json");
        } });

    svr.Post("/transfer", [](const httplib::Request &req, httplib::Response &res)
             {
        try {
            auto body = nlohmann_json::parse(req.body);
            uint64_t senderPublicKey = body["senderPublicKey"];
            uint64_t receiverPublicKey = body["receiverPublicKey"];
            double Amount = body["Amount"];

            {
                std::lock_guard<std::mutex> lock_user(user_mutex);
                std::lock_guard<std::mutex> lock_block(blockchain_mutex);

                for (auto &u : users) {
                    if (u.publicKey == senderPublicKey) {
                        u.transfer(B, senderPublicKey, receiverPublicKey, Amount);
                        break;
                    }
                }
            }

            nlohmann_json reply = {{"status", "transferred"}};
            res.set_header("Connection", "keep-alive");
            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_content(reply.dump(), "application/json");
        } catch (const std::exception &e) {
            res.status = 400;
            res.set_content(string("{\"error\":\"") + e.what() + "\"}", "application/json");
        } });

    // --- Get blockchain endpoint ---
    svr.Get("/chain", [](const httplib::Request &, httplib::Response &res)
            {
        nlohmann_json chain_json = nlohmann_json::array();

        {
            std::lock_guard<std::mutex> lock(blockchain_mutex);
            for (auto &block : B.chain) {
                chain_json.push_back({
                    {"index", block.index},
                    {"username", to_string(block.Username)},
                    {"amount", block.amount},
                    {"hash", block.Hash},
                    {"prevHash", block.previousHash}
                });
            }
        }

        res.set_header("Connection", "keep-alive");
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_content(chain_json.dump(4), "application/json"); });

    // check balance
    svr.Get("/checkbalance", [](const httplib::Request &req, httplib::Response &res)
            {
    try {
        if (!req.has_param("publicKey")) {
            res.status = 400;
            res.set_content("{\"error\":\"Missing publicKey parameter\"}", "application/json");
            return;
        }

        string publicKey = req.get_param_value("publicKey");
        double balance = 0.0;

        {
            std::lock_guard<std::mutex> lock_user(user_mutex);
            std::lock_guard<std::mutex> lock_block(blockchain_mutex);

            for (auto &u : users) {
                if (to_string(u.publicKey) == publicKey) {
                    balance = u.getBalance(B);
                    break;
                }
            }
        }

        nlohmann_json reply = {
            {"publicKey", publicKey},
            {"balance", balance}
        };

        res.set_header("Connection", "keep-alive");
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_content(reply.dump(4), "application/json");
    } 
    catch (const std::exception &e) {
        res.status = 400;
        res.set_content(
            nlohmann_json({{"error", e.what()}}).dump(),
            "application/json"
        );
    } });

    cout << "🚀 Blockchain server running at http://localhost:8080" << endl;

    // Start the HTTP server
    if (!svr.listen("0.0.0.0", 8080))
    {
        cerr << "❌ Failed to start server (port may be in use)" << endl;
        return -1;
    }

    return 0;
}

File: streamlit.py
import streamlit as st
import requests

BASE_URL = "http://localhost:8080"  # your C++ blockchain server

st.set_page_config(page_title="Blockchain CLI", layout="centered")

# --- Persistent session state (acts like variables in your C++ loop)
if "users" not in st.session_state:
    st.session_state.users = []
if "current_user" not in st.session_state:
    st.session_state.current_user = None
if "show_wallet_menu" not in st.session_state:
    st.session_state.show_wallet_menu = False

def is_float(s):
    try:
        float(s)
        return True
    except ValueError:
        return False #check for float. will use later
    
# --- Main Menu
st.title("💰 Simple Blockchain Interface")

if not st.session_state.show_wallet_menu:
    st.subheader("User Menu")
    choice = st.radio(
        "Choose an option:",
        ["Create new wallet", "Select existing wallet", "View blockchain"],
        horizontal=False,
    )

    if choice == "Create new wallet":
        username = st.text_input("Enter new username:")
        if st.button("Create Wallet"):
            res = requests.post(f"{BASE_URL}/create_wallet", json={"username": username})
            if res.status_code == 200:
                data = res.json()
                st.success(f"Wallet created for {username}")
                st.code(f"Public Key: {data['publicKey']}\nPrivate Key: {data['privateKey']}")
                st.session_state.users.append(data)
            else:
                st.error("Failed to create wallet")

    elif choice == "Select existing wallet":
        if not st.session_state.users:
            st.warning("No wallets found! Please create one first.")
        else:
            usernames = [u["publicKey"] for u in st.session_state.users]
            selected = st.selectbox("Select a wallet:", usernames)
            private_key = st.text_input("Enter private key:", type="password")

            if st.button("Login"):
                user = next((u for u in st.session_state.users if u["publicKey"] == selected), None)
                if user and user["privateKey"] == private_key:
                    st.session_state.current_user = user
                    st.session_state.show_wallet_menu = True
                else:
                    st.error("Invalid private key!")

    elif choice == "View blockchain":
        res = requests.get(f"{BASE_URL}/chain")
        if res.status_code == 200:
            st.json(res.json())
    else:
        st.error("Failed to load blockchain.")


if st.session_state.show_wallet_menu:
    user = st.session_state.current_user
    st.header(f"🔑 Wallet Menu - {user['publicKey']}")

    choice = st.radio(
        "Select an option:",
        ["View Blockchain", "Mine Block","Auto-Mine Block", "Check Balance", "Transfer", "Back"],
        horizontal=False,
    )

    if choice == "View Blockchain":
        res = requests.get(f"{BASE_URL}/chain")
        if res.status_code == 200:
            st.json(res.json())

    elif choice == "Mine Block":
        #private_key = st.text_input("Enter private key:", type="password")
        pubKey = st.text_input("Enter public key:", type="password")
        prevHash = st.text_input("Enter previous hash of the block:", type="password")
        currHash = st.text_input("Enter current hash of the block:", type="password")
        if st.button("Mine Block"):
            res = requests.post(
                f"{BASE_URL}/mine",
                json={
                    "publicKey": int(pubKey) if pubKey.isdigit() else pubKey,
                    "prevHash": prevHash,
                    "currHash": currHash,
                },
            )

            if res.status_code == 200:
                data = res.json()
                if data.get("status") == "mined":
                    st.success("✅ Block mined successfully!")
                else:
                    st.warning(f"⚠️ Mining failed: {data.get('reason', 'Unknown reason')}")
            else:
                st.error(f"❌ Server error: {res.text}")
    
    elif choice == "Auto-Mine Block":
        pubKey = st.text_input("Enter Public Key:")
        if st.button("Auto-Mine"):
            res = requests.post(f"{BASE_URL}/automine", json={"publicKey": int(pubKey)})
            if res.status_code == 200:
                data = res.json()
                st.success(f"🤖 Auto block mined!\nPrev: {data['prevHash']}\nCurr: {data['currHash']}")
            else:
                st.error(f"Error: {res.text}")
    
    elif choice == "Transfer":
        #private_key = st.text_input("Enter private key:", type="password")
        senderPublicKey = user['publicKey']
        receiverPublicKey = st.text_input("Enter receiver public key:", type="password")
        Amount = st.text_input("Enter amount:", type="password")
        if st.button("Transfer"):
            res = requests.post(
                f"{BASE_URL}/transfer",
                json={
                    "senderPublicKey": int(senderPublicKey) if senderPublicKey.isdigit() else senderPublicKey,
                    "receiverPublicKey": int(receiverPublicKey) if receiverPublicKey.isdigit() else receiverPublicKey,
                    "Amount": float(Amount) if is_float(Amount) else Amount,
                },
            )

            if res.status_code == 200:
                data = res.json()
                if data.get("status") == "transferred":
                    st.success("✅ Transferred successfully!")
                else:
                    st.warning(f"⚠️ Transferring failed: {data.get('reason', 'Unknown reason')}")
            else:
                st.error(f"❌ Server error: {res.text}")

    elif choice == "Check Balance":
        public_key_str = str(user["publicKey"])
        res = requests.get(f"http://localhost:8080/checkbalance?publicKey={public_key_str}")
        st.json(res.json())

    elif choice == "Back":
        st.session_state.show_wallet_menu = False
        st.session_state.current_user = None
